crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -1009,4 +1011,80 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn adding_or_removing_file_on_which_module_resolution_depends_invalidates_previously_successful_query_that_still_succeeds(
+    ) {
+        const TYPESHED: MockedTypeshed = MockedTypeshed {
+            versions: "functools: 3.8-",
+            stdlib_files: &[("functools.pyi", "def update_wrapper(): ...")],
+        };
+
+        let TestCase {
+            mut db,
+            src,
+            stdlib,
+            site_packages,
+            ..
+        } = TestCaseBuilder::new()
+            .with_custom_typeshed(TYPESHED)
+            .with_target_version(TargetVersion::Py38)
+            .build();
+
+        let functools_module_name = ModuleName::new_static("functools").unwrap();
+        let stdlib_functools_path = stdlib.join("functools.pyi");
+
+        let functools_module = resolve_module(&db, functools_module_name.clone()).unwrap();
+        assert_eq!(functools_module.search_path(), stdlib);
+        assert_eq!(
+            Some(functools_module.file()),
+            system_path_to_file(&db, &stdlib_functools_path)
+        );
+
+        // Adding a file to site-packages does not invalidate the query,
+        // since site-packages takes lower priority in the module resolution...
+        let site_packages_functools_path = site_packages.join("functools.py");
+        db.write_file(&site_packages_functools_path, "f: int")
+            .unwrap();
+        let events = db.take_salsa_events();
+        assert_will_not_run_function_query::<resolve_module_query, _, _>(
+            &db,
+            |res| &res.function,
+            &ModuleNameIngredient::new(&db, functools_module_name.clone()),
+            &events,
+        );
```

Comment:
@MichaReiser am I doing this right? It... doesn't pass for me locally, but I'm not sure if that's because there's a bug in the module resolver or (probably more likely?) I'm using the test function incorrectly...

Reply:
You want to call `db.clear_salsa_events` before starting the mutations to remove the events from the previous computation. 

You also want to move the assertion after your new `resolve_module` call (that should be cached). 

Another way to think about it:
* Clear the events that you're not interested in. That's mainly the events from creating a "warm" salsa cache
* Execute whatever query you want to assert if it gets executed (or not)
* Call the assertion method to verify that the query was executed, or not

```rust
				// Adding a file to site-packages does not invalidate the query,
        // since site-packages takes lower priority in the module resolution...
        let site_packages_functools_path = site_packages.join("functools.py");
        db.write_file(&site_packages_functools_path, "f: int")
            .unwrap();
				db.clear_salsa_events();
	        
				let functools_module = resolve_module(&db, functools_module_name.clone()).unwrap();
        assert_eq!(functools_module.search_path(), stdlib);
        assert_eq!(
            Some(functools_module.file()),
            system_path_to_file(&db, &stdlib_functools_path)
        );

        let events = db.take_salsa_events();
        assert_will_not_run_function_query::<resolve_module_query, _, _>(
            &db,
            |res| &res.function,
            &ModuleNameIngredient::new(&db, functools_module_name.clone()),
            &events,
        );
```

Reply:
I see. Is it then worth renaming the function from `assert_will_run_function_query` to `assert_function_query_was_not_run`?

Reply:
Yeah, I think using past tense in the function name makes sense. I should have thought more about it but I was just happy when the generic madness finally compiled :laughing: